### PR TITLE
Feature/presets

### PR DESCRIPTION
**Pull request by drempelg**
Enhanced Presets and Schedules for Thermostat Cluster

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: examples/thermostat/linux/thermostat-schedules-manager.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+onEditCancel(ThermostatMatterScheduleManager * mgr, ThermostatMatterScheduleManager::editType type)
+{
+    ChipLogProgress(Zcl, "ThermstatScheduleManager - onEditCancel %s", type == ThermostatMatterScheduleManager::Presets ? "Presets" : "Schedules");
+}
+
+static void 
+onEditCommit(ThermostatMatterScheduleManager * mgr, ThermostatMatterScheduleManager::editType type)
+{
+    ChipLogProgress(Zcl, "ThermstatScheduleManager - onEditCommit %s", type == ThermostatMatterScheduleManager::Presets ? "Presets" : "Schedules");
+    gsActivePresets = gsEditingPresets;
```
Comment (bzbarsky-apple) [inline examples/thermostat/linux/thermostat-schedules-manager.cpp:28]: This looks pretty broken; how does this support editing presets _and_ schedules at the same time?

Also, why is cancel a no-op, etc?

Please make sure issues are filed on all the bugs here.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 2]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
     <item fieldId="2" name="CoolSetpoint" type="int16s" isNullable="true"/>
   </struct>
 
+  <bitmap name="PresetTypeFeaturesBitmap" type="bitmap8">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:113]: Spec says this is derived from map16, no?

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 3]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <cluster code="0x0201"/>
+    <field name="Automatic" mask="0x01"/>
+    <field name="SupportsNames" mask="0x02"/>
+  </bitmap>
+
+  <enum name="PresetScenarioEnum" type="enum8">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:119]: Spec does not say what this is derived from.  Please make sure a spec issue is filed and referenced here.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 4]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+  <struct name="PresetTypeStruct">
+    <cluster code="0x201"/>
+    <item fieldId="0" name="PresetScenario" type="PresetScenarioEnum"/>
+    <item fieldId="1" name="NumberOfPresets" type="int8u"/>
+    <item fieldId="2" name="PresetTypeFeatures" type="PresetTypeFeaturesBitmap"/>
+  </struct>
+
+  <struct name="PresetStruct">
+    <cluster code="0x201"/>
+    <item fieldId="0" name="PresetHandle" type="octet_string" length="16"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:139]: This needs to be nullable, per spec.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 5]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <item fieldId="1" name="PresetScenario" type="PresetScenarioEnum"/>
+    <item fieldId="2" name="Name" type="char_string" length="64" isNullable="true" default="null"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:141]: This and some of these other fields are optional, per spec, no?

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 6]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <item fieldId="3" name="CoolingSetpoint" type="int16s" default="2600"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:142]: Isn't the right type "temperature"?

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 7]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <item fieldId="4" name="HeatingSetpoint" type="int16s" default="2000"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:143]: "temperature"

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 8]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <item fieldId="5" name="BuiltIn" type="boolean" default="false"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:144]: In the spec this is nullable....

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 9]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+  </struct>
+
+  <struct name="QueuedPresetStruct">
+    <cluster code="0x201"/>
+    <item fieldId="0" name="PresetHandle" type="octet_string" length="16"/>
+    <item fieldId="1" name="TransitionTimestamp" type="epoch_s"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:150]: In the spec these are both nullable, no?  If this is targeting some un-merged PR, the PR description here should _very_ clearly point that out so people don't waste time comparing it to the wrong thing.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 10]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+  </struct>
+
+  <bitmap name="ScheduleTypeFeaturesBitmap" type="bitmap8">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:153]: Please put the types in this file in the same order as the spec.  That would significantly ease review.

Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:153]: Spec has this as map16.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 11]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+    <cluster code="0x0201"/>
+    <field name="SupportsPresets" mask="0x01"/>
+    <field name="SupportsSetpoints" mask="0x02"/>
+    <field name="SupportsNames" mask="0x03"/>
+    <field name="SupportsOff" mask="0x04"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:158]: These mask values are very much not right.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml [hunk 12]
```diff
@@ -106,15 +110,96 @@ limitations under the License.
...
+  </bitmap>
+
+  <struct name="ScheduleTypeStruct">
+    <cluster code="0x201"/>
+    <item fieldId="0" name="SystemMode" type="ThermostatSystemMode"/>
+    <item fieldId="1" name="NumberOfSchedules" type="int8u" default="0"/>
+    <item fieldId="2" name="ScheduleTypeFeatures" type="ScheduleTypeFeaturesBitmap"/>
+  </struct>
+
+  <struct name="ScheduleTransitionStruct">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:168]: This has optional fields....

I am stopping here.  Either this is not implementing the spec I am reading, or it matches the spec so badly that there are literally errors in almost every single thing added....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/thermostat-server/thermostat-server.cpp [hunk 13]
```diff
@@ -141,6 +377,120 @@ CHIP_ERROR ThermostatAttrAccess::Write(const ConcreteDataAttributePath & aPath,
...
+            bool currentlyEditing = false;
+            EmberAfStatus status = PresetsEditable::Get(aPath.mEndpointId, &currentlyEditing);
+            if (status != EMBER_ZCL_STATUS_SUCCESS)
+            {
+                StatusIB statusIB(ToInteractionModelStatus(status));
+                return statusIB.ToChipError();
+            }
+            if (currentlyEditing == false)
+            {
+                StatusIB statusIB(ToInteractionModelStatus(EMBER_ZCL_STATUS_INVALID_IN_STATE));
```
Comment (jmeg-sfy) [inline src/app/clusters/thermostat-server/thermostat-server.cpp:397]: are we still using EMBER_ZCL ?

Comment (bzbarsky-apple) [reply src/app/clusters/thermostat-server/thermostat-server.cpp:396]: `ToInteractionModelStatus(EMBER_ZCL_STATUS_INVALID_IN_STATE)` is just `Status::InvalidInState`, no?

So: we do have some EMBER_ZCL around, but it's completely unnecessary here.

File: examples/thermostat/linux/thermostat-schedules-manager.cpp (hunk 1 above)
Comment (drempelg) [reply examples/thermostat/linux/thermostat-schedules-manager.cpp]: sorry, this all WIP.  Maybe shouldn't have created PR, was more checking to see what breaks with autobuilds.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml (hunk 2 above)
Comment (drempelg) [reply src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:124]: I'll have to regenerate my version of the spec and check.  I made mine a while a go, and have been working off that until something was finalized.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml (hunk 6 above)
Comment (drempelg) [reply src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:153]: that's also the type for local temperature, but it's not been defined in the XML anywhere.  I should bound it like the other places in the xml though.

File: src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml (hunk 9 above)
Comment (drempelg) [reply src/app/zap-templates/zcl/data-model/chip/thermostat-cluster.xml:161]: This was targetting older versions of the spec TBH lol.

File: src/app/clusters/thermostat-server/thermostat-server.cpp (hunk 13 above)
Comment (drempelg) [reply src/app/clusters/thermostat-server/thermostat-server.cpp:396]: You'll have to forgive me not know which error codes to use, I'm just looking at other examples in the SDK and trying to do "when in rome".

bool emberAfThermostatClusterSetpointRaiseLowerCallback(app::CommandHandler * commandObj,
                                                        const app::ConcreteCommandPath & commandPath,
                                                        const Commands::SetpointRaiseLower::DecodableType & commandData)

is using this sort of thing, so staying consistent with the file.

Perhaps SDK cleanup should be a different job, and mixing methods in a file should be frowned on :P

**Comment by hasty** [comment]
This has been superseded by subsequent changes to the spec.

**Pull request closed without merging**